src/main/java/com/aws/greengrass/authorization/WildcardTrie.java

```diff
@@ -0,0 +1,311 @@
+ *   terminal Node.
+ */
+public class WildcardTrie {
+    protected static final String GLOB_WILDCARD = "*";
+    protected static final String MQTT_MULTILEVEL_WILDCARD = "#";
+    protected static final String MQTT_SINGLELEVEL_WILDCARD = "+";
+    protected static final String MQTT_LEVEL_SEPARATOR = "/";
+
+    private boolean isTerminal;
+    private boolean isTerminalLevel;
+    private boolean isWildcard;
+    private boolean isMQTTWildcard;
+    private boolean matchAll;
+    private final Map<String, WildcardTrie> children =
+            new DefaultConcurrentHashMap<>(WildcardTrie::new);
+
+    /**
+     * Add allowed resources for a particular operation.
+     * - A new node is created for every occurrence of a wildcard (*, #, +).
+     * - Only nodes with valid usage of wildcards are marked with isWildcard or isMQTTWildcard.
+     * - Any other characters are grouped together to form a node.
+     * - Just a '*' or '#' creates a Node setting matchAll to true and would match all resources
+     *
+     * @param subject resource pattern
+     */
+    public void add(String subject) {
+        if (subject == null) {
+            return;
+        }
+        if (subject.equals(GLOB_WILDCARD)) {
+            WildcardTrie initial = this.children.get(GLOB_WILDCARD);
+            initial.matchAll = true;
+            initial.isTerminal = true;
+            initial.isWildcard = true;
+            return;
+        }
+        if (subject.equals(MQTT_MULTILEVEL_WILDCARD)) {
+            WildcardTrie initial = this.children.get(MQTT_MULTILEVEL_WILDCARD);
+            initial.matchAll = true;
+            initial.isTerminal = true;
+            initial.isMQTTWildcard = true;
+            return;
+        }
+        if (subject.equals(MQTT_SINGLELEVEL_WILDCARD)) {
+            WildcardTrie initial = this.children.get(MQTT_SINGLELEVEL_WILDCARD);
+            initial.isTerminal = true;
+            initial.isMQTTWildcard = true;
+            return;
+        }
+        if (subject.startsWith(MQTT_SINGLELEVEL_WILDCARD + MQTT_LEVEL_SEPARATOR)) {
+            WildcardTrie initial = this.children.get(MQTT_SINGLELEVEL_WILDCARD);
+            initial.isMQTTWildcard = true;
+            initial.add(subject.substring(1), true);
+        }
+
+        add(subject, true);
+    }
+
+    @SuppressWarnings("PMD.AvoidDeeplyNestedIfStmts")
+    private WildcardTrie add(String subject, boolean isTerminal) {
+        if (subject.isEmpty()) {
+            this.isTerminal |= isTerminal;
+            return this;
+        }
+        int subjectLength = subject.length();
+        WildcardTrie current = this;
+        StringBuilder sb = new StringBuilder(subjectLength);
+        for (int i = 0; i < subjectLength; i++) {
+            char currentChar = subject.charAt(i);
+            // Create separate Nodes for wildcards *, # and +
+            // Also tag them wildcard if its a valid usage
+            if (currentChar == GLOB_WILDCARD.charAt(0)) {
```

Comment:
store the result of `charAt` in a const like I said before. charAt performs a bounds check every time, so we don't want to do this all the time for no reason.